node/src/store.rs

```diff
@@ -1,9 +1,136 @@
+}
+
+#[derive(Error, Debug, PartialEq)]
+pub enum InsertError {
+    // TODO: do we care about the distinction
+    #[error("Hash {0} already exists in store")]
+    HashExists(Hash),
+    #[error("Height {0} already exists in store")]
+    HeightExists(u64),
+}
+
+#[derive(Error, Debug, PartialEq)]
+pub enum ReadError {
+    // TODO: should we roll internal errors into one
+    #[error("Store in inconsistent state, lost head")]
+    LostStoreHead,
+    #[error("Store in inconsistent state, height->hash mapping exists, but hash doesn't")]
+    LostHash,
+    #[error("Failed to convert height from usize to u64, should not happen")]
+    HeadHeightConversionError,
+
+    #[error("Header not found in store")]
+    NotFound,
+}
 
 impl Store {
     pub fn new() -> Self {
-        Store {}
+        Store {
+            headers: DashMap::new(),
+            height_to_hash: DashMap::new(),
+        }
+    }
+
+    pub fn get_head_height(&self) -> usize {
+        self.height_to_hash.len()
+    }
+
+    pub fn append_header(&self, header: ExtendedHeader) -> Result<(), InsertError> {
+        let hash = header.hash();
+        let height = header.height(); // TODO: should Store be responsible for making sure we have
+                                      // all 1..=head headers?
+
+        // lock both maps to ensure consistency
+        // this shouldn't deadlock as long as we don't hold references across awaits if any
+        // https://github.com/xacrimon/dashmap/issues/233
+        let hash_entry = self.headers.entry(hash);
+        let height_entry = self.height_to_hash.entry(height.into());
+
+        if matches!(hash_entry, Entry::Occupied(_)) {
+            return Err(InsertError::HashExists(hash));
+        }
+        if matches!(height_entry, Entry::Occupied(_)) {
+            return Err(InsertError::HeightExists(height.into()));
+        }
+
+        info!("Will insert {hash} at {height}");
+        hash_entry.insert(header);
+        height_entry.insert(hash);
+
+        Ok(())
+    }
+
+    pub fn get_head(&self) -> Result<ExtendedHeader, ReadError> {
+        let head_height = match self.height_to_hash.len() {
+            0 => return Err(ReadError::NotFound),
+            height => match u64::try_from(height) {
+                Ok(h) => h,
+                Err(e) => {
+                    error!("Failed to convert {height} from usize to u64, should not happen: {e}");
+                    return Err(ReadError::HeadHeightConversionError);
+                }
+            },
+        };
+
+        let Some(head_hash) = self.height_to_hash.get(&head_height).map(|v| *v) else {
+            error!("height_to_hash[height_to_hash.len()] not found, store is inconsistent");
+            return Err(ReadError::LostStoreHead);
+        };
+
+        match self.headers.get(&head_hash) {
+            Some(v) => Ok(v.clone()),
+            None => {
+                error!("Header with hash {head_hash} for height {head_height} missing");
```

Comment:
I'm not sure if I like the error logging and then returning of the error. It produces a noise to read code. I suggest to implement the error messages in the error type itself and the caller should log it. Also in this case, instead of giving to the caller the responsibility, you can use [`#[instrument(err)]`](https://docs.rs/tracing/latest/tracing/attr.instrument.html) from tracing crate.

Reply:
Good point, switched to `instrument` suggestion and removed explicit error logs.